### DISCUSSION TITLE
Improved version of replacement scans

### DISCUSSION THREAD

**Discussion by daniel-thom** [duckdb/duckdb#14041] [Ideas]
The replacement scan feature in Python is great. However, it has two attributes I don't like:
1. It inevitably leads to "unused variable" warnings from linters. As shown in this example, most linters will complain that `rel` is unused.
```python
rel = duckdb.read_csv("data.csv")
rel2 = duckdb.sql("SELECT * from rel")
```
2. There is ambiguity about collisions between registered table names and variables, at least for new users. I understand that registered table names win.

Spark added a similar feature in v3.4 and it doesn't have these limitations.
```python
df = spark.createDataFrame([(1, 2), (3, 4)], ["a", "b"])
spark.sql("select * from {df}", df=df).show()
```
Would you consider adding an interface similar to Spark's?

**Comment by soerenwolfers** [comment]
now that the replacement scan doesn't walk up the stack anymore it's trivial to write this sort of interface yourself with a two line wrapper

**Reply by daniel-thom** [reply]
Would you mind providing an example? Thanks.

**Reply by soerenwolfers** [reply]
```python
from typing import Dict, Any
import duckdb
import inspect
import pandas as pd

def my_query(query: str, /, **kwargs: Dict[str, Any]) -> duckdb.DuckDBPyRelation:
    inspect.currentframe().f_locals.update(**kwargs)
    return duckdb.query(query)
my_query("SELECT * FROM a", a=pd.DataFrame({'a': [0, 1]}))
```

**Reply by soerenwolfers** [reply]
(this comment that used to be here was nonsense)

**Reply by soerenwolfers** [reply]
by the way, https://peps.python.org/pep-0667/ makes the solution above work reliably in >=3.13

**Reply by ajfriend** [reply]
@soerenwolfers I'm having trouble getting your solution above to work. I would expect the following to fail:

<img width="628" alt="image" src="https://github.com/user-attachments/assets/89a66770-15f1-45bb-ad48-72aa0aae2716">

It seems like it it still walking up the stack, unless I'm doing something wrong?

**Reply by ajfriend** [reply]
My code:

```python
import duckdb
import inspect
import pandas as pd

def my_query(query, **kwargs):
    inspect.currentframe().f_locals.update(kwargs)
    return duckdb.query(query)

a = pd.DataFrame({'a': [0, 1]})

my_query('select * from a')
```

**Reply by ajfriend** [reply]
And a slightly simpler example. @Tishj shouldn't this fail?

<img width="618" alt="image" src="https://github.com/user-attachments/assets/8b88ef25-5b16-4c82-9503-80563780f0e9">

**Reply by soerenwolfers** [reply]
it doesn't fail because `a` is in-scope within `my_query`. If, at the point where you call `duckdb.query`, you can write `print(a)` and not get an error, then you can also use `a` in your query.

**Reply by ajfriend** [reply]
Hmm. Then what does "the replacement scan doesn't walk up the stack anymore" mean in this context?

**Reply by soerenwolfers** [reply]
You have to define the function from my snippet in a different module than the one you're calling it from. In particular, if you're using a Jupiter notebook, you cannot define the function in the notebook itself,

**Reply by ajfriend** [reply]
Ah ok, thanks. Ooof, I should have seen that sooner :/

Anyway, I'm still wondering if there is a way to restrict the context for the replacement scans to **exactly** a provided dictionary, which is what I think you were asking for in #10191. It seems like that is still not possible, unless I'm misunderstanding?

**Reply by ajfriend** [reply]
Oh, I suppose that is achievable if you put your function in a separate module to block the possibility of any outside names being in scope. Cool. So exactly what you have above, but just in a module of its own.

**Reply by soerenwolfers** [reply]
If you define my snippet in its own module and only have 
```python
import inspect as __inspect__
import duckdb as __duckdb__
def my_query(...)
   ...
```
in there, then thats pretty much what you and I have been asking for, except for those three names. You can probably even go a step further and mangle the `my_query` name within `my_query` itself, but I agree it would be nicer if there was simply an official way to do this.

**Comment by Tishj** [comment]
I might be able to be convinced to add this:
```py
con.sql("select * from my_df", my_df=df).show()
```

Where all non-recognized kwarg parameters are temporarily registered under the name key that was used

**Reply by Tishj** [reply]
It's somewhat similar to <https://github.com/duckdb/duckdb/discussions/10191> but there being precedence in spark helps the case a bit, and not having to fight the linter is also a decent reason

**Reply by daniel-thom** [reply]
That sounds great.

**Reply by soerenwolfers** [reply]
I think this only makes sense if no automatic interpolation happens besides the explicitly passed parameters

**Reply by ajfriend** [reply]
I would find something like this super useful! Either via keyword arguments like in this comment, or through a `context` dictionary as suggested in https://github.com/duckdb/duckdb/discussions/10191

(This is essentially what I was trying to get at with https://github.com/duckdb/duckdb/discussions/14770)

**Comment by NickCrews** [comment]
I think this flavor of feature should be implemented using the new t-string syntax built in to python 3.14. See https://github.com/duckdb/duckdb-python/discussions/370
